src/efftester.ml

```diff
@@ -1175,13 +1316,27 @@ module Shrinker = struct
     | Float -> to_term (LitFloat (Gen.generate1 float.gen))
     | Bool -> to_term (LitBool (Gen.generate1 bool.gen))
     | String -> to_term (LitStr (Gen.generate1 StaticGenerators.string_gen))
-    | List _ | Fun _ | Typevar _ -> None
+    | Option _ | List _ | Fun _ | Typevar _ -> None
   ;;
 
   (* determines whether x occurs free (outside a binding) in the arg. exp *)
   let rec fv x = function
     | Lit _ -> false
     | Variable (_, y) -> x = y
+    | Constructor (_typ, _name, payload_lst) ->
+      let rec has_fv lst =
+        match lst with
+        | [] -> false
+        | trm :: rest -> if fv x trm then true else has_fv rest
+      in
+      has_fv payload_lst
+    | PatternMatch (_typ, match_trm, branches, _eff) ->
+      let rec first_true lst =
+        match lst with
+        | [] -> false
+        | (_, trm) :: rest -> if fv x trm then true else first_true rest
```

Comment:
This implementation of yours is computing an over-approximation of the set of free variables. For example if I want to see if `x` is free in the clause `| Some x -> x`, it will return `true`, while in fact `x` is not free here -- the occurrence of `x` is bound in the pattern.

Notice how in the cases for `Lambda` and `App` there is specific logic to return `false` on `fun x -> x` or `let x = .. in x`. This is easy to implement in their case because they have the variable directly at hand; in pattern-matching clauses, the variables may occur deep within a pattern.

You could implement a function `occurs_in_pat : variable -> pattern -> bool` that checks whether a variable occurs within a pattern (it is not a *free* variable, it is a *bound* variable, so one should not use `fv` in that case). Then one could implement a complete version of this case as follows:

```ocaml
let fv_in_branch x (pat, trm) =
  if occurs_in_pat x pat then false else fv x m in
fv  match_trm || List.exists (fv_in_branch x) branches
```